File: 2025/R/solve_day5_part1.R
solve_day5_part1 <- function(input) {
  split_at <- which(input == "")

  ranges <- input[1:(split_at - 1)]
  ranges <- read.table(
    text = ranges,
    sep  = "-",
    col.names = c("start", "end")
  )

  ids <- input[(split_at + 1L):length(input)]
  ids <- as.numeric(ids)

  fresh_ids <- vapply(
    ids,
    \(id) any(id >= ranges[["start"]] & id <= ranges[["end"]]),
    logical(1)
  )

  sum(fresh_ids)
}

File: 2025/python/solve_day5_part1.py
def solve_day5_part1(text):
    split_at = text.index("\n")

    ranges = text[:split_at]
    ranges = [line.split("-") for line in ranges]
    ranges = [[int(range[0]), int(range[1].rstrip())] for range in ranges]

    ids = text[(split_at+1):]
    ids = [int(id) for id in ids]

    id_count = 0

    for id in ids:
        for range in ranges:
            if id >= range[0] and id <= range[1]:
                id_count += 1
                break

    return(id_count)
